"Do not directly await a Task" should apply only to libraries, not to applications

#### Analyzer package

Microsoft.CodeQuality.Analyzers, version [2.6.0](https://www.nuget.org/packages/Microsoft.CodeQuality.Analyzers/2.6.0)

#### Analyzer

DoNotDirectlyAwaitATask

#### Repro steps

1. `dotnet new console`
2. `dotnet add package Microsoft.CodeQuality.Analyzers`
3. Add `<LangVersion>7.1</LangVersion>` to the csproj.
3. Change Program.cs to:

    ```c#
    using System.Threading.Tasks;

    static class Program
    {
        static async Task Main()
        {
            await Task.Delay(1);
        }
    }
    ```
4. `dotnet build`

#### Expected behavior

No warning. `ConfigureAwait(false)` is useful in a library, but is generally not necessary in an application. [The description of the warning](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeQuality.Analyzers/Microsoft.CodeQuality.Analyzers.md#ca2007-do-not-directly-await-a-task) also only talks about a library.

#### Actual behavior

```
> dotnet build
Microsoft (R) Build Engine version 15.5.180.51428 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 62,24 ms for C:\code\tmp\hwapp\hwapp.csproj.
Program.cs(7,15): warning CA2007: Do not directly await a Task without calling ConfigureAwait [C:\code\tmp\hwapp\hwapp.csproj]
Program.cs(3,14): warning CA1812: Program is an internal class that is apparently never instantiated. If so, remove the code from the assembly. If this class is intended to contain only static members, make it static (Shared in Visual Basic). [C:\code\tmp\hwapp\hwapp.csproj]
  hwapp -> C:\code\tmp\hwapp\bin\Debug\netcoreapp2.0\hwapp.dll

Build succeeded.

Program.cs(7,15): warning CA2007: Do not directly await a Task without calling ConfigureAwait [C:\code\tmp\hwapp\hwapp.csproj]
Program.cs(3,14): warning CA1812: Program is an internal class that is apparently never instantiated. If so, remove the code from the assembly. If this class is intended to contain only static members, make it static (Shared in Visual Basic). [C:\code\tmp\hwapp\hwapp.csproj]
    2 Warning(s)
    0 Error(s)

Time Elapsed 00:00:09.96
```

I think the CA1812 above is also invalid; I reported that in a separate issue: https://github.com/dotnet/roslyn-analyzers/issues/1688.

--- a/docs/Analyzer Configuration.md
+++ b/docs/Analyzer Configuration.md
@@ -33,10 +33,11 @@ Note that this is a temporary workaround that is needed until the dotnet compile
 ## Supported .editorconfig options
 This section documents the list of supported .editorconfig key-value options for CA rules.
 
-TODO: For each option, list the CA rule IDs that can be configured based on it.
-
 ### Analyzed API surface
 Option Name: `api_surface`
+
+Configurable Rules: <To be documented>
+
 Option Values:
 | Option Value | Summary |
 | --- | --- |
@@ -45,9 +46,24 @@ Option Values:
 | `private` | Analyzes private APIs that are only visible within the containing type. |
 | `all` | Analyzes all APIs, regardless of the symbol visibility. |
 
-Users can also provide a comma separated list of above option values. For example, `private, internal` configures analysis of the entire non-public API surface.
+Example: `dotnet_code_quality.api_surface = all`
+
+Users can also provide a comma separated list of above option values. For example, `dotnet_code_quality.api_surface = private, internal` configures analysis of the entire non-public API surface.
+
+### Analyzed output kinds
+Option Name: `output_kind`
+
+Configurable Rules: [CA2007](../src/Microsoft.CodeQuality.Analyzers/Microsoft.CodeQuality.Analyzers.md#ca2007-do-not-directly-await-a-task)
+
+Option Values: One or more fields of enum [Microsoft.CodeAnalysis.CompilationOptions.OutputKind](http://source.roslyn.io/#q=Microsoft.CodeAnalysis.OutputKind) as a comma separated list.
+
+Example: `dotnet_code_quality.CA2007.output_kind = ConsoleApplication, DynamicallyLinkedLibrary`
 
 ### Async void methods
 Option Name: `skip_async_void_methods`
-Option Values: `true` or `false`
+
 Configurable Rules: [CA2007](../src/Microsoft.CodeQuality.Analyzers/Microsoft.CodeQuality.Analyzers.md#ca2007-do-not-directly-await-a-task)
+
+Option Values: `true` or `false`
+
+Example: `dotnet_code_quality.CA2007.skip_async_void_methods = true`

--- a/docs/Analyzer Configuration.md
+++ b/docs/Analyzer Configuration.md
@@ -33,8 +33,10 @@ Note that this is a temporary workaround that is needed until the dotnet compile
 ## Supported .editorconfig options
 This section documents the list of supported .editorconfig key-value options for CA rules.
 
+TODO: For each option, list the CA rule IDs that can be configured based on it.
+
 ### Analyzed API surface
-Option Name:  `api_surface`
+Option Name: `api_surface`
 Option Values:
 | Option Value | Summary |
 | --- | --- |
@@ -44,3 +46,8 @@ Option Values:
 | `all` | Analyzes all APIs, regardless of the symbol visibility. |
 
 Users can also provide a comma separated list of above option values. For example, `private, internal` configures analysis of the entire non-public API surface.
+
+### Async void methods
+Option Name: `skip_async_void_methods`
+Option Values: `true` or `false`
+Configurable Rules: [CA2007](../src/Microsoft.CodeQuality.Analyzers/Microsoft.CodeQuality.Analyzers.md#ca2007-do-not-directly-await-a-task)

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/DoNotDirectlyAwaitATask.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/DoNotDirectlyAwaitATask.cs
@@ -41,6 +41,12 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
 
             analysisContext.RegisterCompilationStartAction(context =>
             {
+                if (!context.Options.GetOutputKindsOption(Rule, context.CancellationToken).Contains(context.Compilation.Options.OutputKind))
+                {
+                    // Configured to skip analysis for the compilation's output kind
+                    return;
+                }
+
                 ImmutableArray<INamedTypeSymbol> taskTypes = GetTaskTypes(context.Compilation);
                 if (taskTypes.Any(t => t == null))
                 {

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/DoNotDirectlyAwaitATask.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/DoNotDirectlyAwaitATask.cs
@@ -50,9 +50,19 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
                 context.RegisterOperationBlockStartAction(operationBlockStartContext =>
                 {
                     if (operationBlockStartContext.OwningSymbol is IMethodSymbol method &&
-                        method.IsAsync &&
-                        !method.ReturnsVoid)
+                        method.IsAsync)
                     {
+                        if (method.ReturnsVoid &&
+                            operationBlockStartContext.Options.GetBoolOptionValue(
+                                optionName: EditorConfigOptionNames.SkipAsyncVoidMethods,
+                                rule: Rule,
+                                defaultValue: false,
+                                cancellationToken: operationBlockStartContext.CancellationToken))
+                        {
+                            // Configured to skip this analysis in async void methods.
+                            return;
+                        }
+
                         operationBlockStartContext.RegisterOperationAction(oc => AnalyzeOperation(oc, taskTypes), OperationKind.Await);
                     }
                 });

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/DoNotDirectlyAwaitATask.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/DoNotDirectlyAwaitATask.cs
@@ -47,7 +47,15 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
                     return;
                 }
 
-                context.RegisterOperationAction(oc => AnalyzeOperation(oc, taskTypes), OperationKind.Await);
+                context.RegisterOperationBlockStartAction(operationBlockStartContext =>
+                {
+                    if (operationBlockStartContext.OwningSymbol is IMethodSymbol method &&
+                        method.IsAsync &&
+                        !method.ReturnsVoid)
+                    {
+                        operationBlockStartContext.RegisterOperationAction(oc => AnalyzeOperation(oc, taskTypes), OperationKind.Await);
+                    }
+                });
             });
         }
 

--- a/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/AvoidEmptyInterfacesTests.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/AvoidEmptyInterfacesTests.cs
@@ -154,7 +154,7 @@ public class C
 {{
     {accessibility} interface I {{ }}
 }}",
-                GetAdditionalFile(editorConfigText),
+                GetEditorConfigAdditionalFile(editorConfigText),
                 CreateCSharpResult(4, 16 + accessibility.Length));
         }
 
@@ -190,7 +190,7 @@ Public Class C
     {accessibility} Interface I
     End Interface
 End Class",
-                GetAdditionalFile(editorConfigText),
+                GetEditorConfigAdditionalFile(editorConfigText),
                 CreateBasicResult(3, 16 + accessibility.Length));
         }
 
@@ -207,7 +207,7 @@ public class C
 {{
     {accessibility} interface I {{ }}
 }}",
-                GetAdditionalFile(editorConfigText));
+                GetEditorConfigAdditionalFile(editorConfigText));
         }
 
         [Theory]
@@ -223,7 +223,7 @@ Public Class C
     {accessibility} Interface I
     End Interface
 End Class",
-                GetAdditionalFile(editorConfigText));
+                GetEditorConfigAdditionalFile(editorConfigText));
         }
 
         private static DiagnosticResult CreateCSharpResult(int line, int col)
@@ -235,8 +235,5 @@ End Class",
         {
             return GetBasicResultAt(line, col, AvoidEmptyInterfacesAnalyzer.RuleId, MicrosoftApiDesignGuidelinesAnalyzersResources.AvoidEmptyInterfacesMessage);
         }
-
-        private static FileAndSource GetAdditionalFile(string source)
-            => new FileAndSource() { Source = source, FilePath = ".editorconfig" };
     }
 }

--- a/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/DoNotDirectlyAwaitATaskTests.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/DoNotDirectlyAwaitATaskTests.cs
@@ -272,6 +272,28 @@ End Class
                 GetBasicResultAt(9, 33));
         }
 
+        [Fact, WorkItem(1953, "https://github.com/dotnet/roslyn-analyzers/issues/1953")]
+        public void CSharpAsyncVoidMethod_NoDiagnostic()
+        {
+            var code = @"
+using System.Threading.Tasks;
+
+public class C
+{
+    private Task t;
+    public async void M()
+    {
+        await M1Async();
+    }
+
+    private async Task M1Async()
+    {
+        await t.ConfigureAwait(false);
+    }
+}";
+            VerifyCSharp(code);
+        }
+        
         private DiagnosticResult GetCSharpResultAt(int line, int column)
         {
             return GetCSharpResultAt(line, column, DoNotDirectlyAwaitATaskAnalyzer.RuleId, MicrosoftApiDesignGuidelinesAnalyzersResources.DoNotDirectlyAwaitATaskMessage);

--- a/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/DoNotDirectlyAwaitATaskTests.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/DoNotDirectlyAwaitATaskTests.cs
@@ -1,5 +1,7 @@
 // Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Testing;
 using Test.Utilities;
@@ -340,6 +342,41 @@ public class C
     }
 }";
             VerifyCSharp(code, GetEditorConfigAdditionalFile(editorConfigText), GetCSharpResultAt(9, 15));
+        }
+
+        [Fact, WorkItem(1953, "https://github.com/dotnet/roslyn-analyzers/issues/1953")]
+        public void CSharpSimpleAwaitTask_AnalyzerOption_OutputKind()
+        {
+            var code = @"
+using System.Threading.Tasks;
+
+public class C
+{
+    public async Task M()
+    {
+        Task t = null;
+        await t;
+    }
+}
+";
+            var compilationOptions = new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary);
+
+            // Verify diagnostic with no editor config options.
+            VerifyCSharp(code, compilationOptions, parseOptions: null, expected: GetCSharpResultAt(9, 15));
+
+            // Verify no diagnostic with editor config options that exclude OutputKind.DynamicallyLinkedLibrary.
+            var editorConfigText = "dotnet_code_quality.output_kind = ConsoleApplication";
+            VerifyCSharp(code, GetEditorConfigAdditionalFile(editorConfigText), compilationOptions, parseOptions: null);
+
+            editorConfigText = "dotnet_code_quality.CA2007.output_kind = ConsoleApplication, WindowsApplication";
+            VerifyCSharp(code, GetEditorConfigAdditionalFile(editorConfigText), compilationOptions, parseOptions: null);
+
+            // Verify diagnostic with editor config options that include OutputKind.DynamicallyLinkedLibrary.
+            editorConfigText = "dotnet_code_quality.output_kind = DynamicallyLinkedLibrary";
+            VerifyCSharp(code, GetEditorConfigAdditionalFile(editorConfigText), compilationOptions, parseOptions: null, expected: GetCSharpResultAt(9, 15));
+
+            editorConfigText = "dotnet_code_quality.CA2007.output_kind = ConsoleApplication, DynamicallyLinkedLibrary";
+            VerifyCSharp(code, GetEditorConfigAdditionalFile(editorConfigText), compilationOptions, parseOptions: null, expected: GetCSharpResultAt(9, 15));
         }
 
         private DiagnosticResult GetCSharpResultAt(int line, int column)

--- a/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/DoNotDirectlyAwaitATaskTests.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/DoNotDirectlyAwaitATaskTests.cs
@@ -273,7 +273,7 @@ End Class
         }
 
         [Fact, WorkItem(1953, "https://github.com/dotnet/roslyn-analyzers/issues/1953")]
-        public void CSharpAsyncVoidMethod_NoDiagnostic()
+        public void CSharpAsyncVoidMethod_Diagnostic()
         {
             var code = @"
 using System.Threading.Tasks;
@@ -291,9 +291,57 @@ public class C
         await t.ConfigureAwait(false);
     }
 }";
-            VerifyCSharp(code);
+            VerifyCSharp(code, GetCSharpResultAt(9, 15));
         }
-        
+
+        [Theory, WorkItem(1953, "https://github.com/dotnet/roslyn-analyzers/issues/1953")]
+        [InlineData("dotnet_code_quality.skip_async_void_methods = true")]
+        [InlineData("dotnet_code_quality.CA2007.skip_async_void_methods = true")]
+        public void CSharpAsyncVoidMethod_AnalyzerOption_NoDiagnostic(string editorConfigText)
+        {
+            var code = @"
+using System.Threading.Tasks;
+
+public class C
+{
+    private Task t;
+    public async void M()
+    {
+        await M1Async();
+    }
+
+    private async Task M1Async()
+    {
+        await t.ConfigureAwait(false);
+    }
+}";
+            VerifyCSharp(code, GetEditorConfigAdditionalFile(editorConfigText));
+        }
+
+        [Theory, WorkItem(1953, "https://github.com/dotnet/roslyn-analyzers/issues/1953")]
+        [InlineData("dotnet_code_quality.skip_async_void_methods = false")]
+        [InlineData("dotnet_code_quality.CA2007.skip_async_void_methods = false")]
+        public void CSharpAsyncVoidMethod_AnalyzerOption_Diagnostic(string editorConfigText)
+        {
+            var code = @"
+using System.Threading.Tasks;
+
+public class C
+{
+    private Task t;
+    public async void M()
+    {
+        await M1Async();
+    }
+
+    private async Task M1Async()
+    {
+        await t.ConfigureAwait(false);
+    }
+}";
+            VerifyCSharp(code, GetEditorConfigAdditionalFile(editorConfigText), GetCSharpResultAt(9, 15));
+        }
+
         private DiagnosticResult GetCSharpResultAt(int line, int column)
         {
             return GetCSharpResultAt(line, column, DoNotDirectlyAwaitATaskAnalyzer.RuleId, MicrosoftApiDesignGuidelinesAnalyzersResources.DoNotDirectlyAwaitATaskMessage);

--- a/src/Test.Utilities/DiagnosticAnalyzerTestBase.cs
+++ b/src/Test.Utilities/DiagnosticAnalyzerTestBase.cs
@@ -250,8 +250,13 @@ namespace Test.Utilities
 
         protected void VerifyCSharp(string source, FileAndSource additionalText, params DiagnosticResult[] expected)
         {
+            VerifyCSharp(source, additionalText, compilationOptions: null, parseOptions: null, expected: expected);
+        }
+
+        protected void VerifyCSharp(string source, FileAndSource additionalText, CompilationOptions compilationOptions, ParseOptions parseOptions, params DiagnosticResult[] expected)
+        {
             var additionalFiles = GetAdditionalTextFiles(additionalText.FilePath, additionalText.Source);
-            Verify(source, LanguageNames.CSharp, GetBasicDiagnosticAnalyzer(), additionalFiles, compilationOptions: null, parseOptions: null, expected: expected);
+            Verify(source, LanguageNames.CSharp, GetBasicDiagnosticAnalyzer(), additionalFiles, compilationOptions, parseOptions, expected);
         }
 
         protected void VerifyBasic(string source, params DiagnosticResult[] expected)
@@ -301,8 +306,13 @@ namespace Test.Utilities
 
         protected void VerifyBasic(string source, FileAndSource additionalText, params DiagnosticResult[] expected)
         {
+            VerifyBasic(source, additionalText, compilationOptions: null, parseOptions: null, expected: expected);
+        }
+
+        protected void VerifyBasic(string source, FileAndSource additionalText, CompilationOptions compilationOptions, ParseOptions parseOptions, params DiagnosticResult[] expected)
+        {
             var additionalFiles = GetAdditionalTextFiles(additionalText.FilePath, additionalText.Source);
-            Verify(source, LanguageNames.VisualBasic, GetBasicDiagnosticAnalyzer(), additionalFiles, compilationOptions: null, parseOptions: null, expected: expected);
+            Verify(source, LanguageNames.VisualBasic, GetBasicDiagnosticAnalyzer(), additionalFiles, compilationOptions, parseOptions, expected);
         }
 
         protected void Verify(string source, string language, DiagnosticAnalyzer analyzer, IEnumerable<TestAdditionalDocument> additionalFiles, CompilationOptions compilationOptions, ParseOptions parseOptions, params DiagnosticResult[] expected)

--- a/src/Test.Utilities/DiagnosticAnalyzerTestBase.cs
+++ b/src/Test.Utilities/DiagnosticAnalyzerTestBase.cs
@@ -525,6 +525,9 @@ namespace Test.Utilities
                             .Select(x => KeyValuePair.Create(x.Id, ReportDiagnostic.Default))
                             .ToImmutableDictionaryOrEmpty()));
         }
+
+        protected static FileAndSource GetEditorConfigAdditionalFile(string source)
+            => new FileAndSource() { Source = source, FilePath = ".editorconfig" };
     }
 
     // Justification for suppression: We are not going to compare FileAndSource objects for equality.

--- a/src/Utilities/Analyzer.Utilities.projitems
+++ b/src/Utilities/Analyzer.Utilities.projitems
@@ -43,6 +43,7 @@
     <Compile Include="$(MSBuildThisFileDirectory)Extensions\SymbolVisibility.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Extensions\SyntaxNodeExtensions.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Extensions\UriExtensions.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)Options\EditorConfigOptionNames.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Options\SymbolVisibilityGroup.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Options\AnalyzerOptionsExtensions.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Options\CategorizedAnalyzerConfigOptions.cs" />

--- a/src/Utilities/Options/AnalyzerOptionsExtensions.cs
+++ b/src/Utilities/Options/AnalyzerOptionsExtensions.cs
@@ -21,9 +21,9 @@ namespace Analyzer.Utilities
             DiagnosticDescriptor rule,
             SymbolVisibilityGroup defaultValue,
             CancellationToken cancellationToken)
-            => options.GetEnumOptionValue("api_surface", rule, defaultValue, cancellationToken);
+            => options.GetEnumOptionValue(EditorConfigOptionNames.ApiSurface, rule, defaultValue, cancellationToken);
 
-        private static TEnum GetEnumOptionValue<TEnum>(
+        public static TEnum GetEnumOptionValue<TEnum>(
             this AnalyzerOptions options,
             string optionName,
             DiagnosticDescriptor rule,
@@ -33,6 +33,17 @@ namespace Analyzer.Utilities
         {
             var analyzerConfigOptions = options.GetOrComputeCategorizedAnalyzerConfigOptions(cancellationToken);
             return analyzerConfigOptions.GetEnumOptionValue(optionName, rule, defaultValue);
+        }
+
+        public static bool GetBoolOptionValue(
+            this AnalyzerOptions options,
+            string optionName,
+            DiagnosticDescriptor rule,
+            bool defaultValue,
+            CancellationToken cancellationToken)
+        {
+            var analyzerConfigOptions = options.GetOrComputeCategorizedAnalyzerConfigOptions(cancellationToken);
+            return analyzerConfigOptions.GetOptionValue(optionName, rule, bool.TryParse, defaultValue);
         }
 
         private static CategorizedAnalyzerConfigOptions GetOrComputeCategorizedAnalyzerConfigOptions(

--- a/src/Utilities/Options/AnalyzerOptionsExtensions.cs
+++ b/src/Utilities/Options/AnalyzerOptionsExtensions.cs
@@ -1,6 +1,8 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Immutable;
+using System.Linq;
 using System.IO;
 using System.Runtime.CompilerServices;
 using System.Threading;
@@ -15,24 +17,62 @@ namespace Analyzer.Utilities
     {
         private static readonly ConditionalWeakTable<AnalyzerOptions, CategorizedAnalyzerConfigOptions> s_cachedOptions
             = new ConditionalWeakTable<AnalyzerOptions, CategorizedAnalyzerConfigOptions>();
+        private static readonly ImmutableHashSet<OutputKind> s_defaultOutputKinds =
+            ImmutableHashSet.CreateRange(Enum.GetValues(typeof(OutputKind)).Cast<OutputKind>());
 
         public static SymbolVisibilityGroup GetSymbolVisibilityGroupOption(
             this AnalyzerOptions options,
             DiagnosticDescriptor rule,
             SymbolVisibilityGroup defaultValue,
             CancellationToken cancellationToken)
-            => options.GetEnumOptionValue(EditorConfigOptionNames.ApiSurface, rule, defaultValue, cancellationToken);
+            => options.GetFlagsEnumOptionValue(EditorConfigOptionNames.ApiSurface, rule, defaultValue, cancellationToken);
 
-        public static TEnum GetEnumOptionValue<TEnum>(
+        public static ImmutableHashSet<OutputKind> GetOutputKindsOption(
+            this AnalyzerOptions options,
+            DiagnosticDescriptor rule,
+            CancellationToken cancellationToken)
+            => options.GetNonFlagsEnumOptionValue(EditorConfigOptionNames.OutputKind, rule, s_defaultOutputKinds, cancellationToken);
+
+        private static TEnum GetFlagsEnumOptionValue<TEnum>(
             this AnalyzerOptions options,
             string optionName,
             DiagnosticDescriptor rule,
             TEnum defaultValue,
             CancellationToken cancellationToken)
-            where TEnum: struct
+            where TEnum : struct
         {
             var analyzerConfigOptions = options.GetOrComputeCategorizedAnalyzerConfigOptions(cancellationToken);
-            return analyzerConfigOptions.GetEnumOptionValue(optionName, rule, defaultValue);
+            return analyzerConfigOptions.GetOptionValue(
+                optionName, rule,
+                tryParseValue: (string value, out TEnum result) => Enum.TryParse(value, ignoreCase: true, result: out result),
+                defaultValue: defaultValue);
+        }
+
+        private static ImmutableHashSet<TEnum> GetNonFlagsEnumOptionValue<TEnum>(
+            this AnalyzerOptions options,
+            string optionName,
+            DiagnosticDescriptor rule,
+            ImmutableHashSet<TEnum> defaultValue,
+            CancellationToken cancellationToken)
+            where TEnum : struct
+        {
+            var analyzerConfigOptions = options.GetOrComputeCategorizedAnalyzerConfigOptions(cancellationToken);
+            return analyzerConfigOptions.GetOptionValue(optionName, rule, TryParseValue, defaultValue);
+
+            bool TryParseValue(string value, out ImmutableHashSet<TEnum> result)
+            {
+                var builder = ImmutableHashSet.CreateBuilder<TEnum>();
+                foreach (var kindStr in value.Split(','))
+                {
+                    if (Enum.TryParse(kindStr, ignoreCase: true, result: out TEnum kind))
+                    {
+                        builder.Add(kind);
+                    }
+                }
+
+                result = builder.ToImmutable();
+                return builder.Count > 0;
+            }
         }
 
         public static bool GetBoolOptionValue(

--- a/src/Utilities/Options/CategorizedAnalyzerConfigOptions.cs
+++ b/src/Utilities/Options/CategorizedAnalyzerConfigOptions.cs
@@ -1,6 +1,7 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Linq;
@@ -35,12 +36,15 @@ namespace Analyzer.Utilities
             ImmutableDictionary<string, string>.Empty,
             ImmutableDictionary<string, ImmutableDictionary<string, string>>.Empty);
 
+        private readonly ConcurrentDictionary<string, object> _computedOptionValuesMap;
+
         private CategorizedAnalyzerConfigOptions(
             ImmutableDictionary<string, string> generalOptions,
             ImmutableDictionary<string, ImmutableDictionary<string, string>> specificOptions)
         {
             GeneralOptions = generalOptions;
             SpecificOptions = specificOptions;
+            _computedOptionValuesMap = new ConcurrentDictionary<string, object>();
         }
 
         public ImmutableDictionary<string, string> GeneralOptions { get; }
@@ -100,49 +104,19 @@ namespace Analyzer.Utilities
             return new CategorizedAnalyzerConfigOptions(generalOptions, specificOptions);
         }
 
-        public TEnum GetEnumOptionValue<TEnum>(string optionName, DiagnosticDescriptor rule, TEnum defaultValue)
-            where TEnum : struct
-        {
-            if (TryGetSpecificOptionValue(rule.Id, out var optionValue) ||
-                TryGetSpecificOptionValue(rule.Category, out optionValue) ||
-                TryGetGeneralOptionValue(out optionValue))
-            {
-                return optionValue;
-            }
-
-            return defaultValue;
-
-            // Local functions.
-            bool TryGetSpecificOptionValue(string specificOptionKey, out TEnum specificOptionValue)
-            {
-                if (SpecificOptions.TryGetValue(specificOptionKey, out var specificRuleOptions) &&
-                    specificRuleOptions.TryGetValue(optionName, out var valueString))
-                {
-                    return TryParse(valueString, out specificOptionValue);
-                }
-
-                specificOptionValue = defaultValue;
-                return false;
-            }
-
-            bool TryGetGeneralOptionValue(out TEnum generalOptionValue)
-            {
-                if (GeneralOptions.TryGetValue(optionName, out var valueString))
-                {
-                    return TryParse(valueString, out generalOptionValue);
-                }
-
-                generalOptionValue = defaultValue;
-                return false;
-            }
-
-            bool TryParse(string value, out TEnum result)
-                => Enum.TryParse(value, ignoreCase: true, result: out result);
-        }
-
         public delegate bool TryParseValue<T>(string value, out T parsedValue);
 
         public T GetOptionValue<T>(string optionName, DiagnosticDescriptor rule, TryParseValue<T> tryParseValue, T defaultValue)
+        {
+            if (ReferenceEquals(this, Empty))
+            {
+                return defaultValue;
+            }
+
+            return (T)_computedOptionValuesMap.GetOrAdd(optionName, _ => ComputeOptionValue(optionName, rule, tryParseValue, defaultValue));
+        }
+
+        private T ComputeOptionValue<T>(string optionName, DiagnosticDescriptor rule, TryParseValue<T> tryParseValue, T defaultValue)
         {
             if (TryGetSpecificOptionValue(rule.Id, out var optionValue) ||
                 TryGetSpecificOptionValue(rule.Category, out optionValue) ||

--- a/src/Utilities/Options/CategorizedAnalyzerConfigOptions.cs
+++ b/src/Utilities/Options/CategorizedAnalyzerConfigOptions.cs
@@ -140,7 +140,9 @@ namespace Analyzer.Utilities
                 => Enum.TryParse(value, ignoreCase: true, result: out result);
         }
 
-        public int GetIntegralOptionValue(string optionName, DiagnosticDescriptor rule, int defaultValue)
+        public delegate bool TryParseValue<T>(string value, out T parsedValue);
+
+        public T GetOptionValue<T>(string optionName, DiagnosticDescriptor rule, TryParseValue<T> tryParseValue, T defaultValue)
         {
             if (TryGetSpecificOptionValue(rule.Id, out var optionValue) ||
                 TryGetSpecificOptionValue(rule.Category, out optionValue) ||
@@ -152,23 +154,23 @@ namespace Analyzer.Utilities
             return defaultValue;
 
             // Local functions.
-            bool TryGetSpecificOptionValue(string specificOptionKey, out int specificOptionValue)
+            bool TryGetSpecificOptionValue(string specificOptionKey, out T specificOptionValue)
             {
                 if (SpecificOptions.TryGetValue(specificOptionKey, out var specificRuleOptions) &&
                     specificRuleOptions.TryGetValue(optionName, out var valueString))
                 {
-                    return int.TryParse(valueString, out specificOptionValue);
+                    return tryParseValue(valueString, out specificOptionValue);
                 }
 
                 specificOptionValue = defaultValue;
                 return false;
             }
 
-            bool TryGetGeneralOptionValue(out int generalOptionValue)
+            bool TryGetGeneralOptionValue(out T generalOptionValue)
             {
                 if (GeneralOptions.TryGetValue(optionName, out var valueString))
                 {
-                    return int.TryParse(valueString, out generalOptionValue);
+                    return tryParseValue(valueString, out generalOptionValue);
                 }
 
                 generalOptionValue = defaultValue;

--- a/src/Utilities/Options/EditorConfigOptionNames.cs
+++ b/src/Utilities/Options/EditorConfigOptionNames.cs
@@ -1,0 +1,25 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Analyzer.Utilities
+{
+    /// <summary>
+    /// Option names to configure analyzer execution through an .editorconfig file.
+    /// </summary>
+    public static class EditorConfigOptionNames
+    {
+        // =============================================================================================================
+        // NOTE: Keep this file in sync with documentation at '<%REPO_ROOT%>\docs\Analyzer Configuration.md'
+        // =============================================================================================================
+
+        /// <summary>
+        /// Option to configure analyzed API surface.
+        /// Allowed option values: One or more fields of flags enum <see cref="SymbolVisibilityGroup"/> as a comma separated list..
+        /// </summary>
+        public const string ApiSurface = "api_surface";
+
+        /// <summary>
+        /// Boolean option to skip analysis of async void methods.
+        /// </summary>
+        public const string SkipAsyncVoidMethods = "skip_async_void_methods";
+    }
+}

--- a/src/Utilities/Options/EditorConfigOptionNames.cs
+++ b/src/Utilities/Options/EditorConfigOptionNames.cs
@@ -13,7 +13,7 @@ namespace Analyzer.Utilities
 
         /// <summary>
         /// Option to configure analyzed API surface.
-        /// Allowed option values: One or more fields of flags enum <see cref="SymbolVisibilityGroup"/> as a comma separated list..
+        /// Allowed option values: One or more fields of flags enum <see cref="SymbolVisibilityGroup"/> as a comma separated list.
         /// </summary>
         public const string ApiSurface = "api_surface";
 
@@ -21,5 +21,11 @@ namespace Analyzer.Utilities
         /// Boolean option to skip analysis of async void methods.
         /// </summary>
         public const string SkipAsyncVoidMethods = "skip_async_void_methods";
+
+        /// <summary>
+        /// Option to configure analyzed output kinds, i.e. <see cref="Microsoft.CodeAnalysis.CompilationOptions.OutputKind"/> of the compilation.
+        /// Allowed option values: One or more fields of <see cref="Microsoft.CodeAnalysis.CompilationOptions.OutputKind"/> as a comma separated list.
+        /// </summary>
+        public const string OutputKind = "output_kind";
     }
 }

